pkg/scheduler/helper.go

```diff
@@ -0,0 +1,110 @@
+package scheduler
+
+import (
+	"encoding/json"
+	"reflect"
+
+	"k8s.io/klog/v2"
+
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	"github.com/karmada-io/karmada/pkg/util"
+)
+
+func weatherResourceBindingNeedSchedule(rb *workv1alpha2.ResourceBinding) (bool, string) {
+	appliedPlacementStr := util.GetLabelValue(rb.Annotations, util.PolicyPlacementAnnotation)
+	if needConsideredPlacementChanged(*rb.Spec.Placement, appliedPlacementStr, rb.Status.SchedulerObservedAffinityName) {
+		return true, string(ReconcileSchedule)
+	}
+	if rb.Spec.Placement.ReplicaScheduling != nil && util.IsBindingReplicasChanged(&rb.Spec, rb.Spec.Placement.ReplicaScheduling) {
+		// binding replicas changed, need reschedule
+		return true, string(ScaleSchedule)
+	}
+	if rb.Spec.Replicas == 0 ||
+		rb.Spec.Placement.ReplicaScheduling == nil ||
+		rb.Spec.Placement.ReplicaScheduling.ReplicaSchedulingType == policyv1alpha1.ReplicaSchedulingTypeDuplicated {
+		// Duplicated resources should always be scheduled. Note: non-workload is considered as duplicated
+		// even if scheduling type is divided.
+		return true, string(ReconcileSchedule)
+	}
+
+	// TODO(dddddai): reschedule bindings on cluster change
+	return false, ""
+}
+
+func needConsideredPlacementChanged(
+	placement policyv1alpha1.Placement,
+	appliedPlacementStr string,
+	schedulerObservingAffinityName string,
+) bool {
+	if appliedPlacementStr == "" {
+		return true
+	}
+
+	appliedPlacement := policyv1alpha1.Placement{}
+	err := json.Unmarshal([]byte(appliedPlacementStr), &appliedPlacement)
+	if err != nil {
+		klog.Errorf("Failed to unmarshal applied placement string: %v", err)
+		return false
+	}
+
+	// first check: entire placement does not change
+	if reflect.DeepEqual(placement, appliedPlacement) {
+		return false
+	}
+
+	// second check: except for ClusterAffinities, the placement has changed
+	if !reflect.DeepEqual(placement.ClusterAffinity, appliedPlacement.ClusterAffinity) ||
+		!reflect.DeepEqual(placement.ClusterTolerations, appliedPlacement.ClusterTolerations) ||
+		!reflect.DeepEqual(placement.SpreadConstraints, appliedPlacement.SpreadConstraints) ||
+		!reflect.DeepEqual(placement.ReplicaScheduling, appliedPlacement.ReplicaScheduling) {
+		return true
+	}
+
+	// third check: check weather ClusterAffinities has changed
+	return checkClusterAffinities(placement.ClusterAffinities, appliedPlacement.ClusterAffinities, schedulerObservingAffinityName)
+}
+
+func checkClusterAffinities(
```

Comment:
One question：
Assume a policy like:
```
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: nginx
spec:
  resourceSelectors:
    - apiVersion: apps/v1
      kind: Deployment
      name: nginx
  placement:
    clusterAffinities:
      - affinityName: dc-shanghai
        clusterNames:
          - unavailable
      - affinityName: dc-beijing
        clusterNames:
          - member1
      - affinityName: dc-hongkong
        clusterNames:
          - member2
```
Now scheduler uses the affinity `dc-beijing`. Then users change the placement to:
```
  placement:
    clusterAffinities:
      - affinityName: dc-shanghai
        clusterNames:
          - unavailable
      - affinityName: dc-hongkong
        clusterNames:
          - member2
      - affinityName: dc-beijing
        clusterNames:
          - member1
```
Do we need to reschedule?

Reply:
No, for affinity `dc-beijing`, it doesn't change.

Reply:
So let's assume the original configuration is:
```yaml
  placement:
    clusterAffinities:
      - affinityName: dc-shanghai
        clusterNames:
          - unavailable
      - affinityName: dc-hongkong
        clusterNames:
          - member2
      - affinityName: dc-beijing
        clusterNames:
          - member1
```
And scheduler choose dc-hongkong, then user update the configuration to:
```yaml
  placement:
    clusterAffinities:
      - affinityName: dc-beijing
        clusterNames:
          - member1
      - affinityName: dc-shanghai
        clusterNames:
          - unavailable
      - affinityName: dc-hongkong
        clusterNames:
          - member2
```
Now if dc-hongkong is not satisfied with the requests, looks like it will return an error and will not try to schedule to dc-beijing and dc-shanghai